lib/cc/service/safe_webhook.rb

```diff
@@ -0,0 +1,54 @@
+module CC
+  class Service
+    class SafeWebhook
+      InvalidWebhookURL = Class.new(StandardError)
+
+      def self.getaddress(host)
+        @resolv ||= Resolv::DNS.new
+        @resolv.getaddress(host)
+      end
+
+      def initialize(url)
+        @url = url
+      end
+
+      # Resolve the Host to an IP address, validate that it doesn't point to
+      # anything internal, then alter the request to be for the IP directly with
+      # an explicit Host header given.
+      #
+      # See http://blog.fanout.io/2014/01/27/how-to-safely-invoke-webhooks/#ip-address-blacklisting
+      def validate!(request)
+        uri = URI.parse(url)
+        address = self.class.getaddress(uri.host)
+
+        if internal?(address)
+          raise_invalid("resolves to a private IP address")
+        end
+
+        alter_request(request, uri, address)
+      rescue URI::InvalidURIError, Resolv::ResolvError => ex
+        raise_invalid(ex.message)
+      end
+
+      private
+
+      attr_reader :url
+
+      def internal?(address)
+        address.start_with?("127.") ||
```

Comment:
Yeah, it's a TODO in the PR description to finalizer this. that wikipedia page or some RFC should give us something comprehensive to use.